File: typescript/packages/node/src/commands/builtin/redis/csplit.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { ResourceName, command, specOf, csplitGeneric } from '@struktoai/mirage-core'
import type { RedisAccessor } from '../../../accessor/redis.ts'
import { stream as redisStream } from '../../../core/redis/stream.ts'
import { writeBytes as redisWrite } from '../../../core/redis/write.ts'

export const REDIS_CSPLIT = command({
  name: 'csplit',
  resource: ResourceName.REDIS,
  spec: specOf('csplit'),
  fn: (accessor: RedisAccessor, paths, texts, opts) =>
    csplitGeneric(
      paths,
      texts,
      opts,
      (p) => redisStream(accessor, p),
      (p, d) => redisWrite(accessor, p, d),
    ),
  write: true,
})

File: python/mirage/commands/builtin/generic_bind/builders/csplit.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from functools import partial

from mirage.accessor.base import Accessor
from mirage.commands.builtin.generic.csplit import csplit as generic_csplit
from mirage.commands.builtin.generic_bind.adapter import (Builder, CommandIO,
                                                          Operation, bound_op,
                                                          resolve_or_empty)
from mirage.commands.config import CommandOpts
from mirage.commands.spec import SPECS
from mirage.commands.spec.flag_view import FlagView
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


async def csplit(ops: CommandIO, accessor: Accessor, paths: list[PathSpec],
                 texts: list[str],
                 opts: CommandOpts) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(opts.flags, spec=SPECS["csplit"])
    paths = await resolve_or_empty(ops, accessor, paths, opts.index)
    prefix_flag = fl.raw("prefix")
    prefix = prefix_flag if isinstance(prefix_flag, (str, PathSpec)) else "xx"
    return await generic_csplit(
        paths,
        texts,
        read_bytes=bound_op(ops.read_bytes, accessor, opts.index),
        write_bytes=partial(ops.require(Operation.WRITE), accessor),
        stdin=opts.stdin,
        prefix=prefix,
        mount_prefix=opts.mount_prefix,
        digits=int(fl.as_str("digits") or "2"),
        suffix_format=fl.as_str("suffix_format"),
        keep_on_error=fl.as_bool("keep_files"),
        silent=fl.as_bool("quiet") or fl.as_bool("silent"),
        suppress_matched=fl.as_bool("suppress_matched"),
        elide_empty=fl.as_bool("elide_empty_files"))


BUILDER = Builder('csplit', csplit, write=True)
